test/501_migrations/normal_migrations.t

```diff
@@ -0,0 +1,53 @@
+The 501 parsetree contains a parsing modificacion.
+[compare_on.exe <file>] checks if there's a diff between the
+AST's resulting from
+1. parsing <file> on 5.1.0 directly
+2. parsing <file> on 5.1.0, migrating down to 5.0.0 and migrating back to 5.1.0
+We only expect a diff in one special case.
+
+  $ echo "let x : int = 5" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let _ : int = 5" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let f : type a b c. a -> b -> c = fun x y -> assert false" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let f = (fun (type a) (type b) (type c) -> (fun x y -> assert false : a -> b -> c))" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let _ = (fun (type a) (type b) (type c) -> (fun x y -> assert false : a -> b -> c))" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let f : type a . a -> a = fun x -> x" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let (x, y) : (int * int) = assert false" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let f: type a. a option -> _ = assert false" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let f : 'a . 'a = (fun (type a) -> (assert false : a))" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+Here we expect a diff (downwards migrating should yield the same as in the example right above).
+However, something is wrong.
+  $ echo "let f : type a . a = assert false" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
```

Comment:
For this case, what happens is that on the surface language the `down` transform is injective:
```ocaml
let f: type a.a = assert false
```
is mapped onto
```ocaml
let (f: 'a . 'a) = (fun (type a) -> assert false : a)
```
but the type annotation in the pattern `(f:'a . 'a)` is not allowed in the surface language.
Which means that in order to observe a diff in the roundtrip, one need to start with ppx-constructed AST in 5.1
```ocaml
let (f: 'a . 'a) = (fun (type a) -> assert false : a)
```
which will end up being normalized into
```ocaml
let f: type a. a = assert false
```
by the roundtrip. I am not sure if we want to test those cases, however?

Reply:
> on the surface language the `down` transform is injective

I've double-checked again and for me,
```
let f : 'a . 'a = (fun (type a) -> (assert false : a))
```
is accepted syntax and is parsed into the same AST in `5.0` (up to locations) as
```
let f : type a . a = assert false
```
Those two examples are parsed into different ASTs in `5.1`. Am I misunderstanding something? If not, isn't this an example showing that also on the surface language `down` is not injective?

Reply:
First, yes, I agree that
``ocaml
let f : 'a . 'a = (fun (type a) -> (assert false : a))
```
and 
```ocamll
let f : type a . a = assert false
```
are parsed to the same node in 5.0 . I was mislead by the absence of diff and the problem with  ppx built parsetree exists that should ignore for now.

What's really at play here is that the `down` function is using a new encoding for the 5.1 term
```ocaml
let f : 'a . 'a = (fun (type a) -> (assert false : a))
```
which is transformed into 5.0
```ocaml
let (f: ø . ('a . 'a)) = ((fun (type a) -> (assert false : a)) : ('a . 'a))
```
because we are in the 
```
| ( Some (Pvc_constraint { locally_abstract_univars; typ }),
        { Ast_500.Parsetree.ppat_desc = Ppat_var _; ppat_attributes = [] } )
```
branch (with `locally_abstract_univars=[]`).
Then the `up` function pick this term as a pattern `let (x: ... . typ ) = (exp : typ)` and 
transform it back to 5.1
```ocaml
let f : 'a . 'a = (fun (type a) -> (assert false : a))
```
A good way to test this issue is to assert in the `up` translation that the `typ` term in  `(pat: ... . typ)` is not a `Ptyp_poly` itself with
```diff
--- a/astlib/migrate_500_501.ml
+++ b/astlib/migrate_500_501.ml
@@ -287,6 +287,7 @@ and copy_value_binding :
              { ptyp_desc = Ptyp_poly (args_tyvars, rt) } );
        ppat_attributes = [];
       } ->
+          assert (match rt.ptyp_desc with Ptyp_poly _ ->  false | _ -> true );
           Some (pat, args_tyvars, rt, `Var)
       | {
        Ast_500.Parsetree.ppat_desc = Ppat_constraint (pat, rt);
```
which currently fails on the `normal_migrations.t` test.

Reply:
I see, that makes a lot of sense! Are you currently fixing this as well?

Reply:
Yep, I should have a fix ready soon.